examples/step-67b/step-67b.cc

```diff
@@ -0,0 +1,1712 @@
+  {
+    VectorizedArrayType result;
+    for (unsigned int v = 0; v < VectorizedArrayType::size(); ++v)
+      {
+        Point<dim> p;
+        for (unsigned int d = 0; d < dim; ++d)
+          p[d] = p_vectorized[d][v];
+        result[v] = function.value(p, component);
+      }
+    return result;
+  }
+
+
+  template <int dim, typename VectorizedArrayType, int n_components = dim + 2>
+  Tensor<1, n_components, VectorizedArrayType>
+  evaluate_function(const Function<dim> &                  function,
+                    const Point<dim, VectorizedArrayType> &p_vectorized)
+  {
+    AssertDimension(function.n_components, n_components);
+    Tensor<1, n_components, VectorizedArrayType> result;
+    for (unsigned int v = 0; v < VectorizedArrayType::size(); ++v)
+      {
+        Point<dim> p;
+        for (unsigned int d = 0; d < dim; ++d)
+          p[d] = p_vectorized[d][v];
+        for (unsigned int d = 0; d < n_components; ++d)
+          result[d][v] = function.value(p, d);
+      }
+    return result;
+  }
+
+
+  // @sect3{MPI-3.0 shared memory}
+
+  // The following code snippet shows a class storing an MPI sub-communicator.
+  // It furthermore handles the creation
+  // and freeing of the sub-communicator. The user can specify the size of
+  // the sub-communicator. If the size is set to -1, all MPI processes of a
+  // shared-memory domain are combined to a group. The specified size is
+  // decisive for the benefit of the shared-memory capabilities of MatrixFree
+  // and, therefore, setting the <code>size</code> to <code>-1</code> is a
+  // reasonable choice. By setting, the size to <code>1</code> users explicity
+  // disable the MPI-3.0 shared-memory features of MatrixFree and rely
+  // completely on MPI-2.0 features, like <code>MPI_Isend</code> and
+  // <code>MPI_Irecv</code>.
+  class SubCommunicatorWrapper
+  {
+  public:
+    SubCommunicatorWrapper(
+      const MPI_Comm &   comm,
+      const unsigned int group_size = numbers::invalid_unsigned_int)
+    {
+#if defined(DEAL_II_WITH_MPI) && DEAL_II_MPI_VERSION_GTE(3, 0)
+      if (group_size == 1)
+        {
+          this->comm = MPI_COMM_SELF;
+        }
+      else if (group_size != numbers::invalid_unsigned_int)
+        {
+          const auto rank = Utilities::MPI::this_mpi_process(comm);
+
+          MPI_Comm_split(comm, rank / group_size, rank, &this->comm);
+        }
+      else
+        {
+          const auto rank = Utilities::MPI::this_mpi_process(comm);
+
+          MPI_Comm_split_type(
+            comm, MPI_COMM_TYPE_SHARED, rank, MPI_INFO_NULL, &this->comm);
+        }
+#else
+      (void)comm;
+      (void)group_size;
+      this->comm = MPI_COMM_SELF;
+#endif
+    }
+
+    ~SubCommunicatorWrapper()
+    {
+#ifdef DEAL_II_WITH_MPI
+      if (this->comm != MPI_COMM_SELF)
+        MPI_Comm_free(&comm);
+#endif
+    }
+
+    const MPI_Comm &get_communicator() const
+    {
+      return comm;
+    }
+
+  private:
+    MPI_Comm comm;
+  };
```

Comment:
FYI @tjhei

Reply:
Awesome, I need that. Do you mind if we pull this out into the main library? Do you want to prep the PR or do you want me to do that?

Reply:
Feel free to do so. I would maybe drop the second option (`MPI_Comm_split(comm, rank / group_size, rank, &this->comm);`) because it makes the assumption that ranks in the communicator have not be reordered before that.